man/httppower.8
.TH PowerMan 8 "Release 2.3" "LLNL" "PowerMan"
.SH NAME
httppower \- communicate with HTTP based power distribution units
.SH SYNOPSIS
.B httppower
.I "[--url URL]"
.LP
.SH DESCRIPTION
.B httppower
is a helper program for
.B powerman
which enables it to communicate with HTTP based power distribution units.
It is run interactively by the powerman daemon.
.SH OPTIONS
.TP
.I "-u, --url URL"
Set the base URL.
.SH INTERACTIVE COMMANDS
The following commands are available at the httppower> prompt:
.TP
.I "auth user:pass"
Authenticate to the base URL with specified user and password,
using ``basic'' HTTP authentication which sends the user and password
over the network in plain text.
.TP
.I "seturl URL"
Set the base URL.  Overrides the command line option.
.TP
.I "get [URL-suffix]"
Send an HTTP GET to the base URL with the optional URL-suffix appended.
.TP
.I "post [URL-suffix] key=val[&key=val]..."
Send an HTTP POST to the base URL with the optional URL-suffix appended,
and key-value pairs as argument.
.SH "FILES"
/usr/sbin/httppower
.br
/usr/sbin/powermand
.br
/etc/powerman/powerman.conf
.SH "SEE ALSO"
powerman(1) powerman.conf(5) powermand(8)


NEWS
=========================================================================
Release Notes for PowerMan version 2.3                        12 Nov 2008
=========================================================================

* Make init script work on Solaris.

* IPv6 support.

* Include example powerman.conf file (/etc/powerman/powerman.conf.example)

* Added support for Cyclades PM20, PM42.

=========================================================================
Release Notes for PowerMan version 2.2                        27 Aug 2008
=========================================================================

* Send a SIGINT to coprocesses rather than just closing file
  descriptors during powermand shutdown.  This is required to 
  terminate an ssh client running in coprocess mode.

* Make it possible to only define the _ranged version of scripts.  Scripts 
  are selected using the following precedence: _all, _ranged, singlet.

* Drop "soft power status" support.  This allowed iceboxes to detect 
  whether a node was powered up or not, independently of plug state.

* [ipmipower] Drop _all and singlet version of scripts. 

* [ilom] Revised Sun Integrated Lights Out Management support 
  to work over ssh and serial.  Dropped shared console/serial 
  "ilom-inline" support.

* [icebox3] Now supports both v3 and v4 Ice Box.  Users of icebox4
  should change their device types to icebox3 in powerman.conf.

* [plmpower] Added support for controlling Insteon/X10 devices via 
  SmartLabs PLM 2412S.

* [hp3488] Added support for modular GPIB-based HP 3488A switch/control 
  unit via gpib-utils project.

* [ics8064] Added support for VXI-11-based ICS 8064 16-port relay unit
  via gpib-utils project.

* [powerman] Improved efficiency of powerman-controlling-powerman 
  configurations, when one powerman controls a subset of the plugs of 
  another powerman through the use of _ranged scripts.

* Enhanced integrated test suite.
  Note: tests all pass on AIX now.

=========================================================================
Release Notes for PowerMan version 2.1                        18 Jun 2008
=========================================================================

* Client overhaul.

* [apcpdu3] Fix intermittent query failures.

* Fixed off by one bug in server that allowed some delays in dev 
  scripts to take longer than programmed.

* Use raw mode in pipe (|&) connections to avoid local tty echo.

* Run server in /var/run/powerman rather than /etc/powerman.
  Create this directory if it doesn't exist.

* Enhanced integrated test suite.
  Note: all tests do not currently pass on AIX.

=========================================================================
Release Notes for PowerMan version 2.0                        01 Jun 2008
=========================================================================

* Autoconf/automake integration.

* Integrated test suite.

* Portability to Solaris, AIX, and OS/X.

* Support for selecting power control targets using genders (-g option).

* Cleanup and refactoring.

=========================================================================
Release Notes for PowerMan version 1.0.32                     24 Jan 2008
=========================================================================

* Support for new apc firmware (apcpdu3.dev) [Trent D'Hooge]

* Added httppower utility for interfacing to power controllers that are
  exclusively web-based.

* Support for Digital Loggers, Inc. power controllers.

* Support for APPRO power controller.

=========================================================================
Release Notes for PowerMan version 1.0.31                     20 Oct 2007
=========================================================================

(Includes changes in 1.0.29 and 1.0.30)

* Support SUN ILOM inline with serial console, i.e.
  device "ilom25" "ilom-inline" "/usr/bin/conman -Q -j ilc25 |&"

* Handle config files with embedded carriage returns [Trent D'Hooge
  and Todd Heer].

* Ipmipower device script tuning [Al Chu].

* Minor build/packaging tweaks for building under mock, etc.

=========================================================================
Release Notes for PowerMan version 1.0.28                     07 May 2007
=========================================================================

* Increase "cycle" delays from 2s to 4s on all devices that implement
  cycle as an explicit on,delay,off.

=========================================================================
Release Notes for PowerMan version 1.0.27                     24 Apr 2007
=========================================================================

* Add heartbeat stonith module.

* Fix bug in baytech-rpc28-nc device support that affected plugs > 10.

=========================================================================
Release Notes for PowerMan version 1.0.26                     21 Dec 2006
=========================================================================

* Support 8-port APC AP7900 and likely AP7901, AP7920, and AP7921.
  [Martin Petersen]

* Work around a bug that causes baytech rpc3's to resturn some plugs
  as status "unknown" on login timeout/reconnect.

* Add ranged power control support for faster power control w/ ipmipower.

=========================================================================
Release Notes for PowerMan version 1.0.25                     16 Aug 2006
=========================================================================

* Support newer baytech RPC-3 firmware.

* Support icebox v4 [Jarod Wilson]

=========================================================================
Release Notes for PowerMan version 1.0.24                     30 May 2006
=========================================================================

* Telnet state machine now works with Digi terminal server in telnet 
  mode and logs ignored telnet option requests.

* Several powerman.conf examples are now included in the RPM doc area.


src/httppower.c
/*****************************************************************************\
 *  $Id:$
 *****************************************************************************
 *  Copyright (C) 2007-2008 The Regents of the University of California.
 *  Produced at Lawrence Livermore National Laboratory (cf, DISCLAIMER).
 *  Written by Jim Garlick <garlick@llnl.gov>
 *  UCRL-CODE-2002-008.
 *  
 *  This file is part of PowerMan, a remote power management program.
 *  For details, see <http://www.llnl.gov/linux/powerman/>.
 *  
 *  PowerMan is free software; you can redistribute it and/or modify it under
 *  the terms of the GNU General Public License as published by the Free
 *  Software Foundation; either version 2 of the License, or (at your option)
 *  any later version.
 *  
 *  PowerMan is distributed in the hope that it will be useful, but WITHOUT 
 *  ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or 
 *  FITNESS FOR A PARTICULAR PURPOSE.  See the GNU General Public License 
 *  for more details.
 *  
 *  You should have received a copy of the GNU General Public License along
 *  with PowerMan; if not, write to the Free Software Foundation, Inc.,
 *  59 Temple Place, Suite 330, Boston, MA  02111-1307  USA.
\*****************************************************************************/

#if HAVE_CONFIG_H
#include "config.h"
#endif
#include <stdio.h>
#include <libgen.h>
#if HAVE_CURL
#include <curl/curl.h>
#else
#error httppower needs curl support!
#endif
#include <getopt.h>
#include <string.h>
#include <stdlib.h>

#include "xtypes.h"
#include "xmalloc.h"
#include "error.h"
#include "argv.h"

static char *url = NULL;
static char *userpwd = NULL;
static char errbuf[CURL_ERROR_SIZE];

#define OPTIONS "u:"
static struct option longopts[] = {
        {"url", required_argument, 0, 'u' },
        {0,0,0,0},
};

void help(void)
{
    printf("Valid commands are:\n");
    printf("  auth user:passwd\n");
    printf("  seturl url\n");
    printf("  get [url]\n");
    printf("  post [url] key=val[&key=val]...\n");
}

char *
_make_url(char *str)
{
    char *myurl = NULL;

    if (str && url) {
        myurl = xmalloc(strlen(url) + strlen(str) + 2);
        sprintf(myurl, "%s/%s", url, str);
    } else if (str && !url) {
        myurl = xstrdup(str);
    } else if (!str && url) {
        myurl = xstrdup(url);
    }
    return myurl;
}

void post(CURL *h, char **av)
{
    char *myurl = _make_url(av[0]);
    char *postdata = av[1] ? xstrdup(av[1]) : NULL;

    if (postdata && myurl) {
        curl_easy_setopt(h, CURLOPT_URL, myurl);
        curl_easy_setopt(h, CURLOPT_POSTFIELDS, postdata);
        if (curl_easy_perform(h) != 0)
            printf("Error: %s\n", errbuf);
        curl_easy_setopt(h, CURLOPT_URL, "");
        curl_easy_setopt(h, CURLOPT_POSTFIELDS, "");
    } else
        printf("Nothing to post!\n");

    if (myurl)
        xfree(myurl);
    if (postdata)
        xfree(postdata);
}

void get(CURL *h, char **av)
{
    char *myurl = _make_url(av[0]);

    if (myurl) {
        curl_easy_setopt(h, CURLOPT_URL, myurl);
        if (curl_easy_perform(h) != 0)
            printf("Error: %s\n", errbuf);
        curl_easy_setopt(h, CURLOPT_URL, "");
    } else 
        printf("Nothing to get!\n");

    if (myurl)
        xfree(myurl);
}

void seturl(CURL *h, char **av)
{
    if (av[0] == NULL) {
        printf("Usage: seturl http://...\n");
        return;
    }
    if (url)
        xfree(url);
    url = xstrdup(av[0]);
}

void auth(CURL *h, char **av)
{
    if (av[0] == NULL) {
        printf("Usage: auth user:passwd\n");
        return;
    }
    if (userpwd)
        xfree(userpwd);
    userpwd = xstrdup(av[0]);
    curl_easy_setopt(h, CURLOPT_USERPWD, userpwd);
    curl_easy_setopt(h, CURLOPT_HTTPAUTH, CURLAUTH_BASIC);
}

int docmd(CURL *h, char **av)
{
    int rc = 0;

    if (av[0] != NULL) {
        if (strcmp(av[0], "help") == 0)
            help();
        else if (strcmp(av[0], "quit") == 0)
            rc = 1;
        else if (strcmp(av[0], "auth") == 0)
            auth(h, av + 1);
        else if (strcmp(av[0], "seturl") == 0)
            seturl(h, av + 1);
        else if (strcmp(av[0], "get") == 0)
            get(h, av + 1);
        else if (strcmp(av[0], "post") == 0)
            post(h, av + 1);
        else 
            printf("type \"help\" for a list of commands\n");
    }
    return rc;
}

void shell(CURL *h)
{
    char buf[128];
    char **av;
    int rc = 0;

    while (rc == 0) {
        printf("httppower> ");
        fflush(stdout);
        if (fgets(buf, sizeof(buf), stdin)) {
            av = argv_create(buf, "");
            rc = docmd(h, av);
            argv_destroy(av);
        } else
            rc = 1;
    }
}

void
usage(void)
{
    fprintf(stderr, "Usage: httppower [--url URL]\n");
    exit(1);
}

int
main(int argc, char *argv[])
{
    CURL *h;
    int c;

    err_init(basename(argv[0]));

    while ((c = getopt_long(argc, argv, OPTIONS, longopts, NULL)) != EOF) {
        switch (c) {
            case 'u': /* --url */
                url = xstrdup(optarg);
                break;
            default:
                usage();
                break;
        }
    }
    if (optind < argc)
        usage();

    if (curl_global_init(CURL_GLOBAL_ALL) != 0)
        err_exit(FALSE, "curl_global_init failed");
    if ((h = curl_easy_init()) == NULL)
        err_exit(FALSE, "curl_easy_init failed");

    curl_easy_setopt(h, CURLOPT_TIMEOUT, 5);
    curl_easy_setopt(h, CURLOPT_ERRORBUFFER, errbuf);
    curl_easy_setopt(h, CURLOPT_FAILONERROR, 1);

    shell(h);

    curl_easy_cleanup(h);    	
    if (userpwd)
        xfree(userpwd);
    if (url)
        xfree(url);
    exit(0);
}

/*
 * vi:tabstop=4 shiftwidth=4 expandtab
 */
